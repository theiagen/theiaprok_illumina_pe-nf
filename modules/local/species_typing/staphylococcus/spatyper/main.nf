process SPATYPER {
    tag "$meta.id"
    label 'process_low'

    container "us-docker.pkg.dev/general-theiagen/biocontainers/spatyper:0.3.3--pyhdfd78af_3"

    input:
    tuple val(meta), path(assembly)
    val do_enrich 

    output:
    tuple val(meta), path("*.tsv"), emit: tsv
    tuple val(meta), path("*_values.txt"), emit: spatyper_value_results
    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def enrich = do_enrich ? "--do_enrich" : ""
    """
    # get versioning
    spaTyper --version 2>&1 | sed 's/^.*spaTyper //' | tee VERSION
    
    spaTyper \\
      ${enrich} \\
      --fasta ${assembly} \\
      --output ${prefix}.tsv

    python3 <<EOF
    import csv

    TYPE = []
    REPEATS = []

    with open("./${prefix}.tsv",'r') as tsv_file:
        tsv_reader=csv.reader(tsv_file, delimiter="\t")
        next(tsv_reader, None)  # skip the headers
        for row in tsv_reader:
            TYPE.append(row[-1])
            REPEATS.append(row[-2])

        with open ("TYPE_values.txt", 'wt') as TYPE_fh:
            TYPE_fh.write(','.join(TYPE))

        with open ("REPEATS_values.txt", 'wt') as REPEATS_fh:
            REPEATS_fh.write(','.join(REPEATS))
    EOF

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        spaTyper: \$(VERSION)
        python3: \$(python3 --version | cut -d ' ' -f 2)
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}.tsv
    echo "No type found" > TYPE_values.txt
    echo "No repeats found" > REPEATS_values.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        spaTyper: \$(VERSION)
        python3: \$(python3 --version | cut -d ' ' -f 2)
    END_VERSIONS
    """
}
