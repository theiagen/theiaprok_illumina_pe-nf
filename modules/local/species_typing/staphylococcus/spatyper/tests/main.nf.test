nextflow_process {
    name "Test Process SPATYPER"
    script "../main.nf"
    process "SPATYPER"
    
    test("spatyper test") {
        when {
            process {
                """
                input[0] = Channel.of([
                    [ id:'test_sample' ],
                    file("$projectDir/bin/SRR7706353_filtered_contigs.fasta", checkIfExists: true)
                ])
                input[1] = false
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out.tsv,
                    process.out.repeats,
                    process.out.type,
                    process.out.versions
                    ).match() }
            )
        }
    }
    
    test("spatyper - stub") {
        options "-stub"
        when {
            process {
                """
                input[0] = Channel.of([
                    [ id:'test_sample' ],
                    file("$projectDir/bin/SRR7706353_filtered_contigs.fasta", checkIfExists: true)
                ])
                input[1] = false  // Set do_enrich to false for variation
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}
