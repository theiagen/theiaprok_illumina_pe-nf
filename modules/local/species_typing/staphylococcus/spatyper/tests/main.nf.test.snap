{
    "spatyper test": {
        "content": [
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "test_sample.tsv:md5,10a31ab23a232ce70a56c72134953fb7"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "REPEATS:md5,238e5efb505a6f5785895073a675d035"
                ]
            ],
            [
                [
                    {
                        "id": "test_sample"
                    },
                    "TYPE:md5,573f89c6dac8854674a872d0091e6a5a"
                ]
            ],
            null
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.8"
        },
        "timestamp": "2025-06-25T20:18:42.885581414"
    },
    "spatyper - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "REPEATS:md5,358427305e412fd00ab05251f491dfa8"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "TYPE:md5,0104a1bf80547cbd6bc99eb00df49a5d"
                    ]
                ],
                "repeats": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "REPEATS:md5,358427305e412fd00ab05251f491dfa8"
                    ]
                ],
                "tsv": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "test_sample.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "type": [
                    [
                        {
                            "id": "test_sample"
                        },
                        "TYPE:md5,0104a1bf80547cbd6bc99eb00df49a5d"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.8"
        },
        "timestamp": "2025-06-25T20:18:47.473943345"
    }
}